csrc/dynamic_transform.cpp

```diff
@@ -20,43 +20,253 @@
+    }
+  }
+
+  const auto& getInfo() const {
+    return info_;
+  }
 
+ private:
   using IterVisitor::handle;
 
-  // Analyze a dynamic reshape and generate AnalyzeViewResult
-  void handle(ViewOp* op) override;
+  //! Find views that have symbolic outputs
+  void handle(ViewOp* op) override {
+    auto inp_tv = op->in()->as<TensorView>();
+    auto out_tv = op->out()->as<TensorView>();
+    // If there's no symbolic axis, this is a static reshape op
+    if (out_tv->domain()->hasSymbolicAxis()) {
+      info_.dynamic_reshapes_.push_back(op);
+
+      // Input and output extent expressions both affect concretization
+      const auto& inp_dom =
+          TensorDomain::noReductions(inp_tv->getMaybeRFactorDomain());
+      for (const auto id : inp_dom) {
+        // Try and evaluate the extent so that intermediate expressions are
+        // cached in expr_eval_
+        auto ext = info_.expr_eval_.evaluate(id->extent());
+        if (!ext.has_value()) {
+          leaf_dynamic_vals_.push_back(id->extent());
+        }
+      }
+      const auto& out_dom = out_tv->getMaybeRFactorDomain();
+      for (const auto id : out_dom) {
+        auto ext = info_.expr_eval_.evaluate(id->extent());
+        if (!ext.has_value()) {
+          leaf_dynamic_vals_.push_back(id->extent());
+        }
+      }
+    }
+  }
 
-  // We handle IterDomain "Resize" ops at TensorView level
-  void handle(TensorView* tv) override;
+  //! Detect dynamic IterDomain transforms when handling TensorViews
+  void handle(TensorView* tv) override {
+    const auto& rfd = tv->getMaybeRFactorDomain();
+    for (auto id : rfd) {
+      if (!id->definition() || id->getIterType() != IterType::Symbolic) {
+        continue;
+      }
+      if (auto op = dynamic_cast<Resize*>(id->definition())) {
+        info_.dynamic_resizes_.push_back(op);
+        // extent of output determines its IterType
+        leaf_dynamic_vals_.push_back(id->extent());
+        // warm up extent evaluation
+        info_.expr_eval_.evaluate(id->extent());
+      }
+    }
+  }
 
-  const auto& getInfo() const {
-    return info_;
+  //! Process vector of leaf dynamic values by finding inputs and recording the
+  //! result into info_
+  void finalizeDynamicVals() {
+    const auto inputs = InputsOf::outputs(info_.fusion(), leaf_dynamic_vals_);
+    info_.root_dynamic_vals_.insert(inputs.begin(), inputs.end());
   }
 
  private:
-  ExpressionEvaluator* expr_eval_ = nullptr;
+  DynamicTransformInitialInfo info_;
 
-  DynamicTransformConcretizationInfo info_;
+  std::vector<Val*> leaf_dynamic_vals_;
 };
 
-DynamicTransformInfoBuilder::DynamicTransformInfoBuilder(
-    Fusion* fusion,
-    ExpressionEvaluator* expr_eval)
-    : expr_eval_(expr_eval), info_(fusion) {
-  TORCH_INTERNAL_ASSERT(
-      !fusion->isA<kir::Kernel>(),
-      "Invalid container. Kernel container not allowed.\n");
+void DynamicTransformConcretizationInfo::analyzeReshapes(
```

Comment:
Just to confirm, is this just moved? Same for `analyzeResizes`.

Reply:
Yes, these were previously in the `DynamicTransformInfoBuilder::handle()` overrides. I just copied them to here and update to use an explicit `ExpressionEvaluator`. I don't keep `expr_eval` inside `DynamicTransformConcretizationInfo` since I use it with `std::hash` so I'd like to keep it small. Instead I copy the `ExpressionEvaluator` from the initial info in `getConcretizationInfo` and pass it to these functions.